_posts/24-07-2022-what_is_a_reader_and_why_it_matters.md

```diff
@@ -0,0 +1,134 @@
+category: Scala
+description: Without mentioning monads!
+---
+### The dream
+Imagine we have functions:
+
+```Scala
+def f: A => B = ???
+def g: B => C = ???
+def h: C => D = ???
+```
+
+The implementation may be whatever we want it to be, these functions might parse some data, perform calculations, format strings et cetera. Now, often we want to chain multiple operations together - ie. parse a JSON with sales data, then extract monthly sales figures and finally output an average. After all, coding is all about gluing blocks together to create a useful program.
+We could, of course, write `h(g(f(x)))` each time we want the whole functionality, but if it's used frequently, giving it a proper name is a good idea.
+Scala gives us many different ways of gluing functions together, but I'm going to stick to the `andThen` from `Function1` trait - because it behaves almost like the mathematical `∘` operator, just in reverse order. With it creating a function `i: A => D` is as simple as writing:
+
+```Scala
+def i: A => D = f andThen g andThen h
+```
+
+### The reality
+This kind of composition, obviously, works only for functions with one argument - with multi-argument functions we quickly run into a world of trouble. Quite unfortunate, since the vast majority of those we encounter on a daily basis need more than one input. But what if all of our functions look like this?:
+
+```Scala
+def f: E => A => B = ???
+def g: E => B => C = ???
+def h: E => C => D = ???
+```
+
+This is again a common thing - oftentimes we need to pass some sort of a context (or other dependency) to our method. It would be useful to be able to compose them, even if just to save us some writing.
+They share the type of the first argument - it must have made our task easier. What do we need to do to be able to compose them in a sensible way?
+
+## The promise
+First of all, let's simplify things a bit and imagine a function that magically reads some value `V` from an environment `E`. It's going to have a type `E => V`. We're going to call it `Reader` for precisely that reason, and we're going to wrap it in a case class for convenience:
+
+```Scala
+case class Reader[E, V](read: E => V)
+```
+
+We're halfway there, we just need to notice that function parameters should be swappable (with possible minor tweaks to the logic inside - but usually we don't depend on their order anyway). Then we can rewrite our `f`, `g`, and `h` as:
+
+```Scala
+def f: A => E => B = ???
+def g: B => E => C = ???
+def h: C => E => D = ???
+```
+
+There's a lot of `Reader`s just waiting in line. We can now rewrite our functions again:
+
+```Scala
+def f: A => Reader[E, B] = ???
+def g: B => Reader[E, C] = ???
+def h: C => Reader[E, D] = ???
+```
+
+### The solution
+Now we just need to compose these somehow. First of all, let's give our Reader a helper method:
+
+```Scala
+case class Reader[E, V](read: E => V) {
+  def pipeTo[W](reader: V => Reader[E, W]): Reader[E, W] = Reader { env =>
```

Comment:
Hmmm maybe for clarity
```suggestion
  def pipeTo[W](other: V => Reader[E, W]): Reader[E, W] = Reader { env =>
```
Coz the use of `read` and `reader` is pretty confusing

Reply:
Done.